lib/als/async-hook.js

```diff
@@ -0,0 +1,150 @@
+import asyncHooks from 'async_hooks';
+import { getActiveEvent, getInfo } from './als';
+import { EventType } from '../constants';
+import { Ntp } from '../ntp';
+import { checkForNested } from '../tracer/tracer';
+import { stackTrace } from '../utils';
+
+const getResource = (asyncId) => {
+  if (!asyncId) return;
+
+  const info = getInfo();
+
+  if (!info) return;
+
+  return info.resources.get(asyncId);
+};
+
+export const isDescendantOf = (asyncId, parentAsyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return false;
+
+  if (resource.triggerAsyncId === parentAsyncId) return true;
+
+  const parent = getResource(resource.triggerAsyncId);
+
+  if (!parent) return false;
+
+  return isDescendantOf(parent.asyncId, parentAsyncId);
+};
+
+export const getAncestryChainOfIds = (asyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return [];
+
+  if (!resource.triggerAsyncId) return [asyncId];
+
+  const parent = getResource(resource.triggerAsyncId);
+
+  if (!parent) return [asyncId];
+
+  return [...getAncestryChainOfIds(parent.asyncId), asyncId];
+};
+
+export const getActiveEventResource = () => {
+  const activeEvent = getActiveEvent();
+
+  return getResource(activeEvent?.asyncId);
+};
+
+const captureStartTime = (asyncId) => {
+  const resource = getResource(asyncId);
+
+  if (!resource) return;
+
+  resource.startTime = Ntp._now();
+};
+
+const captureEndTime = (asyncId) => {
+  const resource = getResource(asyncId);
+  if (!resource) return;
+
+  // Some resources do not trigger before/after hooks.
+  if (!resource.startTime) {
+    resource.startTime = resource.initAt;
+  }
+
+  resource.endTime = Ntp._now();
+  resource.duration = resource.endTime - resource.startTime;
+
+  const info = getInfo();
+
+  // We can ignore async events which take less than a millisecond for performance.
+  if (resource.duration === 0) return;
+  if (!info) return;
+
+  // Some events finish after the trace is already processed.
+  if (info.trace.isEventsProcessed) return;
+
+  const activeEvent = getActiveEvent();
+
+  const asyncEvent = {
+    type: EventType.Async,
+    at: resource.startTime,
+    endAt: resource.endTime,
+    asyncId,
+    level: activeEvent?.level + 1 || 0,
+  };
+
+  if (activeEvent && !activeEvent.endAt) {
+    checkForNested(activeEvent);
+    activeEvent.nested.push(asyncEvent);
+  } else {
+    info.trace.events.push(asyncEvent);
+  }
+};
+
+export const AsyncResourceType = {
+  PROMISE: 'PROMISE',
+};
+
+const hook = asyncHooks.createHook({
+  init (asyncId, type, triggerAsyncId) {
+    const info = getInfo();
+
+    if (!info) return;
+
+    // We don't want to capture anything other than promise resources.
+    if (type !== AsyncResourceType.PROMISE) return;
+
+    const activeEvent = getActiveEvent();
```

Comment:
Would it be worth checking the level, and skipping this work if the event would be nested too deep?

Reply:
Yes, definitely